include/linker/devicetree_regions_mpu.h

```diff
@@ -0,0 +1,49 @@
+/*
+ * Copyright (c) 2021, Carlo Caione <ccaione@baylibre.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <linker/devicetree_regions.h>
+
+/*
+ * List of compatibles whose YAML files are including mem-region.yaml to
+ * support the 'zephyr,memory-region-mpu' property.
+ */
+#define _DT_COMPATIBLES mmio_sram, arm_dtcm, arm_itcm, nxp_imx_dtcm, nxp_imx_itcm,	\
+			st_stm32_ccm, st_stm32_backup_sram
+
+#define _CHECK_ATTR_FN(node_id, FN, ...)						\
+	COND_CODE_1(UTIL_AND(DT_NODE_HAS_PROP(node_id, zephyr_memory_region_mpu),	\
+			     DT_NODE_HAS_PROP(node_id, zephyr_memory_region)),		\
+		   (FN(node_id, __VA_ARGS__)),						\
+		   ())
+
+#define _EXPAND_MPU_FN(node_id, FN, ...)		\
+	FN(LINKER_DT_NODE_REGION_NAME(node_id),		\
+	   DT_REG_ADDR(node_id),			\
+	   DT_REG_SIZE(node_id),			\
+	   DT_STRING_TOKEN(node_id, zephyr_memory_region_mpu))
+
+#define _CHECK_APPLY_FN(compat, FN, ...)					\
+	DT_FOREACH_STATUS_OKAY_VARGS(compat, _CHECK_ATTR_FN, FN, __VA_ARGS__)
+
+#define _DT_SRAM_FOREACH(compat, mpu_fn) _CHECK_APPLY_FN(compat, _EXPAND_MPU_FN, mpu_fn)
+
+/*
+ * Helper macro to apply an MPU_FN macro to all the SRAM regions declared using
+ * the 'zephyr,memory-region-mpu' property.
+ *
+ * 'mpu_fn' must take the form:
+ *
+ *   #define MPU_FN(name, base, size, attr) ...
+ *
+ * Example:
+ *
+ *   static const struct arm_mpu_region mpu_regions[] = {
+ *       ...
+ *       DT_SRAM_MPU(MPU_FN)
+ *       ...
+ *   };
+ */
+#define DT_SRAM_MPU(mpu_fn) FOR_EACH_FIXED_ARG(_DT_SRAM_FOREACH, (), mpu_fn, _DT_COMPATIBLES)
```

Comment:
note: MPU attributes are also applicable to e.g. S**D**RAM (example case: STM32H7)

Reply:
I'll call it 'DT_REGION_MPU`